lib/semantics/assignment.cpp

```diff
@@ -29,194 +29,79 @@ using namespace Fortran::parser::literals;
-};
-
-struct ForallContext {
-  explicit ForallContext(const ForallContext *that) : outer{that} {}
-
-  const ForallContext *outer{nullptr};
-  std::optional<parser::CharBlock> constructName;
-  std::vector<Control> control;
-  std::optional<MaskExpr> maskExpr;
-  std::set<parser::CharBlock> activeNames;
-};
-
-struct WhereContext {
-  WhereContext(MaskExpr &&x, const WhereContext *o, const ForallContext *f)
-    : outer{o}, forall{f}, thisMaskExpr{std::move(x)} {}
-  const WhereContext *outer{nullptr};
-  const ForallContext *forall{nullptr};  // innermost enclosing FORALL
-  std::optional<parser::CharBlock> constructName;
-  MaskExpr thisMaskExpr;  // independent of outer WHERE, if any
-  MaskExpr cumulativeMaskExpr{thisMaskExpr};
-};
-
 class AssignmentContext {
 public:
-  explicit AssignmentContext(SemanticsContext &c) : context_{c} {}
-  AssignmentContext(const AssignmentContext &c, WhereContext &w)
-    : context_{c.context_}, where_{&w} {}
-  AssignmentContext(const AssignmentContext &c, ForallContext &f)
-    : context_{c.context_}, forall_{&f} {}
-
+  explicit AssignmentContext(SemanticsContext &context) : context_{context} {}
+  AssignmentContext(AssignmentContext &&) = default;
+  AssignmentContext(const AssignmentContext &) = delete;
   bool operator==(const AssignmentContext &x) const { return this == &x; }
 
+  template<typename A> void PushWhereContext(const A &);
+  void PopWhereContext();
+  void PushForallContext() { ++forallDepth_; }
+  void PopForallContext();
   void Analyze(const parser::AssignmentStmt &);
   void Analyze(const parser::PointerAssignmentStmt &);
-  void Analyze(const parser::WhereStmt &);
-  void Analyze(const parser::WhereConstruct &);
-  void Analyze(const parser::ForallConstruct &);
-
-  template<typename A> void Analyze(const parser::UnlabeledStatement<A> &stmt) {
-    context_.set_location(stmt.source);
-    Analyze(stmt.statement);
-  }
-  template<typename A> void Analyze(const common::Indirection<A> &x) {
-    Analyze(x.value());
-  }
-  template<typename A> std::enable_if_t<UnionTrait<A>> Analyze(const A &x) {
-    std::visit([&](const auto &y) { Analyze(y); }, x.u);
-  }
-  template<typename A> void Analyze(const std::list<A> &list) {
-    for (const auto &elem : list) {
-      Analyze(elem);
-    }
-  }
-  template<typename A> void Analyze(const std::optional<A> &x) {
-    if (x) {
-      Analyze(*x);
-    }
-  }
+  void Analyze(const parser::ConcurrentControl &);
 
 private:
-  void Analyze(const parser::WhereConstruct::MaskedElsewhere &);
-  void Analyze(const parser::MaskedElsewhereStmt &);
-  void Analyze(const parser::WhereConstruct::Elsewhere &);
-
   void CheckForPureContext(const SomeExpr &lhs, const SomeExpr &rhs,
       parser::CharBlock rhsSource, bool isPointerAssignment);
-
-  MaskExpr GetMask(const parser::LogicalExpr &, bool defaultValue = true);
-
+  void CheckShape(parser::CharBlock, const SomeExpr *);
+  void CheckForallIndexesUsed(parser::CharBlock, const SomeExpr &);
```

Comment:
"indices" is the preferred plural of "index" in mathematical contexts.

Reply:
I will change it if you have a strong preference. I chose "Indexes" because irregular plurals make it harder to search the code for terms like "index".

Reply:
"index" is not incorrect, so leave it alone if that's your preference.